codeU5.py

```diff
@@ -0,0 +1,127 @@
+# -*- coding: utf-8 -*-
+from collections import defaultdict
+import numpy as np
+import unittest
+
+'''Given a dictionary (a list of words in lexicographic order) of all words in an unknown/invented
+language, find the alphabet (an ordered list of characters) of that language.
+This language can contain any character (of the native char data type). Upper/lower case
+characters are treated differently for simplicity. Assume standard lexicographical ordering
+(order by characters from left to right, if X is a prefix of Y then X is sorted before Y), just with
+an unknown order of characters.
+Write a function that will receive an ordered list of strings, and returns an ordered list of
+characters'''
+
+class Alphabet:
+    def __init__(self,word_list):
+        
+        '''Args:
+            word_list: non-empty list of unique words from the given language
+        '''
+        
+        self.words = word_list
+        self.partial_order = [] # a list of partial order lists [char1,char2], char1 < char2
+        self.all_chars = set() # unordered set of all characters
+        self.sorted_alphabet = [] # the list of characters in the alphabetical order
+        for word in self.words:
+            self.all_chars.update(set(word))
+        self.neigh_dict = defaultdict(list) # dictionary of char: [list of chars going after the given char] 
+        
+    def find_common_prefix(self, word1, word2):
+        
+        if len(word1) < len(word2):
+            word_shorter = word1
+            word_longer = word2
+        else:
+            word_longer = word1
+            word_shorter = word2
+            
+        for i, c in enumerate(word_shorter):
+            if c != word_longer[i]:
+                return word_shorter[:i]
+    
+    def find_partial_order(self, word1, word2):
+        
+        '''Returns:
+            partial order [char_prev,char_next]
+            inferred from two words word1 and word2
+        '''
+        
+        prefix = self.find_common_prefix(word1, word2)
+        char_first = word1[len(prefix)]
+        char_next = word2[len(prefix)]
+        return [char_first, char_next]

```

Comment:
IDK if this matters at all in practice, but generally multiple return values are returned as a tuple `(char_first, char_next)` rather than a `list`.

I'm not sure if there is a particular reason to prefer one or the other. I was wondering if tuples could be more efficien to construct but there doesn't seem to be any obvious performance difference at all:

```
>>> timeit.timeit('a, b = (1, 2)', number=long(1e8))
1.7991349697113037
>>> timeit.timeit('[a, b] = [1, 2]', number=long(1e8))
1.7812578678131104
```

One difference is that tuples (of hashable types) are hashable while lists are not:

```
>>> s = set()
>>> s.add((1, 2))
>>> s.add([1, 2])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unhashable type: 'list'
```